File: java/java-features-trainer/resources/learnProjects/java/LearnProject/src/warehouse/Warehouse.java
package warehouse;

import java.util.HashMap;
import java.util.Map;

import util.Utils;

public final class Warehouse {
    // Fruit name to amount of it in warehouse
    private final Map<String, Integer> entry = new HashMap<>(); // Apple, banana, etc...

    public Warehouse() {
        String[] availableFruits = Utils.FRUITS;
        for (String fruit : availableFruits) {
            entry.put(fruit, 0);
        }
    }

  /**
   * @param fruitName some fruit name from Utils.FRUITS (mango, apple...)
   */
    public void addFruits(String fruitName, int quantity) {
        Integer curQuantity = entry.get(fruitName);
        if (curQuantity != null) {
            entry.put(fruitName, curQuantity + quantity);
        }
        else {
            throw new IllegalArgumentException("Not found fruit with name: " + fruitName);
        }
    }

    public boolean takeFruit(String fruitName) {
        Integer curQuantity = entry.get(fruitName);
        if (curQuantity == null) {
            throw new IllegalArgumentException("Not found fruit with name: " + fruitName);
        }
        else if (curQuantity > 0) {
            entry.put(fruitName, curQuantity - 1);
            return true;
        }
        return false;
    }

    public void printAllFruits() {
        for (Map.Entry<String, Integer> pair : entry.entrySet()) {
            System.out.println(pair.getKey() + ": " + pair.getValue());
        }
    }
}

File: plugins/kotlin/features-trainer/resources/learnProjects/kotlin/KotlinLearningProject/src/warehouse/Warehouse.kt
package warehouse

import util.Utils

class Warehouse {
    // Fruit name to amount of it in warehouse
    private val entry: MutableMap<String, Int> = HashMap() // Apple, banana, etc...

    init {
        val availableFruits: Array<String> = Utils.FRUITS
        for (fruit in availableFruits) {
            entry[fruit] = 0
        }
    }

    /**
     * @param fruitName some fruit name from [Utils.FRUITS] (mango, apple...)
     */
    fun addFruits(fruitName: String, quantity: Int) {
        val curQuantity = entry[fruitName]
        if (curQuantity != null) {
            entry[fruitName] = curQuantity + quantity
        } else {
            throw IllegalArgumentException("Not found fruit with name: $fruitName")
        }
    }

    fun takeFruit(fruitName: String): Boolean {
        val curQuantity = entry[fruitName]
        requireNotNull(curQuantity) { "Not found fruit with name: $fruitName" }
        if (curQuantity > 0) {
            entry[fruitName] = curQuantity - 1
            return true
        }
        return false
    }

    fun printAllFruits() {
        for ((key, value) in entry) {
            println("$key: $value")
        }
    }
}